### PR TITLE
Stampy updates

### PR DESCRIPTION
**Pull request by stampycode**
Added a variety of improvements, including addition of PATCH and HEAD HTTP methods, header case sensitivity fixes, and fixes for this.merge() unexpectedly modifying input params.

### CODE REVIEW

File: Tester.js [hunk 1]
```diff
@@ -213,13 +145,23 @@ Tester.prototype = {
       this.logger.debug('with body: ' + JSON.stringify(this.req.body));
     }
 
-    this.request({url: url, method: method, body: this.req.body, headers: headers, json: json}, function (error, response, body) {
+    var sendObj = {
+      url: url,
+      method: method,
+      body: this.req.body,
+      headers: headers
+      //json: (headers['Content-Type'] === 'application/json')
```
Comment (vatson) [inline Tester.js]: Please remove the code that isn't used

File: Tester.js [hunk 2]
```diff
@@ -239,6 +181,17 @@ Tester.prototype = {
     next();
   },
 
+  modifyAndAssertJSON: function(custom_assertion) {
```
Comment (vatson) [inline Tester.js:189]: Could you explain what it is and for what (an example from the life)?

Comment (stampycode) [inline Tester.js:189]: This is for applying an assertion in the JSON with a custom callback for modifying the JSON before the comparison takes place, so to account for dynamic response values, like ID fields, for example.

```
this.Then(/^the response should match the following, taking into account that "([^"]*)" are dynamic fields:$/, function(fields, exp_json, callback) {
    fields = fields.split(',');
    this.api.modifyAndAssertJSON(function(actual_json, call){
            for (i=0; i < fields.length; i++) {
                var field_name = fields[i];
                actual_json[field_name] = "";
            }
            call(exp_json, actual_json);
        })
        .then(callback);
});
```

Comment (vatson) [inline Tester.js:189]: Oh, that sounds reasonable. But it's still a bit confused. Need to add more detailed descriptions and examples in the documentation

File: Tester.js [hunk 3]
```diff
@@ -239,6 +180,38 @@ Tester.prototype = {
...
+   *       for (i = 0; i < fields.length; i++) {
+   *         field_name = fields[i];
+   *         actual_json[field_name] = "";
+   *       }
+   *       call(exp_json, actual_json);
+   *     })
+   *     .then(callback);
+   *   }
+   * );
+   */
```
Comment (vatson) [inline Tester.js]: Nah. I meant to update README.md ([API Reference section](https://github.com/linkshare/plus.garden.api#api-reference)) with this description  cuz the added method is the part of public api.

Could you move it to README.md and add an example with feature's step that is implemented in your step definition?

Comment (stampycode) [inline Tester.js]: I've added it, and a usage example to README.md, do you want me to remove the usage comment from the code?

Comment (vatson) [inline Tester.js]: Yeah. Please remove it from comment block. Let's keep the code in the same style (the rest methods are not commented).

Comment (stampycode) [inline Tester.js]: comments removed.

**Comment by vatson** [comment]
Thanks. Really nice job :+1:

**Pull request merged**